### PR TITLE
Mainframe Data Access Framework

### PR DESCRIPTION
**Pull request by gitizenme**
Jeff,

For this review focus on Mainframe_Data_Access_Framework.md - this is an initial version and there will be more updates as I discover the moving parts of FDR and the tech stack. 

Joe

### CODE REVIEW

File: architecture/Mainframe_Data_Access_Framework.md [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+ 
+
+### Aspects
+
+| **Structural Aspect**    | **Artifact​**          | **Notes**                                           |
+|--------------------------|-----------------------|-----------------------------------------------------|
+| Reference Client         | Versioned source code | REPL shell                                          |
+| API Layer                | Versioned object code | HTTPS REST/JSON + OAuth,  JWT                       |
+| Service Layer            | Versioned object code | Microservice(s)                                     |
```
[the remark this answers is not in the document]
Comment (gitizenme) [reply architecture/Mainframe_Data_Access_Framework.md]: An API may orchestrate calls to multiple services as part of completing a request from a client. So, I guess it should be _Services Layer_ .

File: architecture/Mainframe_Data_Access_Framework.md [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+### Monitoring
+
+Monitoring for the DAF is TBD. However, in general the following common
+activities should be reported:
+
+-   Authentication - success (normal), failed (elevated)
+
+-   Exceptions (elevated)
+
```
[the remark this answers is not in the document]
Comment (gitizenme) [reply architecture/Mainframe_Data_Access_Framework.md]: I agree, my original thought was to have that as part of the Performance section but I think calling it out here is a better approach.

File: architecture/Mainframe_Data_Access_Framework.md [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+The DAF is provided to application development engineering teams as a standalone
+component which identified by a semantic version number. There are 4 primary
+environments that are considered for ready for deployment of the DAF.
+
+-   Engineering (-engineering)
+
+-   QA (-qa)
+
+-   Staging (-staging)
```
[the remark this answers is not in the document]
Comment (gitizenme) [reply architecture/Mainframe_Data_Access_Framework.md]: Probably not, my thinking here is that these were _logical_ environments that are static and are always sourced from the Production container when starting a new feature or fixing a bug. I don't think I expressed that very well based on your reading. I'll fix that up...

File: architecture/Mainframe_Data_Access_Framework.md [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+
+-   Production (-prod)
+
+The semantic version for each framework instance is identified by the
+environment name appended to the semantic version. An application client is
+paired with the instance of the DAF that it requires in order to implement one
+or more features. Shown the diagram below are several configurations of an
+application client and the DAF framework.
+
+![Environment Diagram](diagrams/Mainframe_Data_Access_Framework-Environment_Diagram.svg)
```
[the remark this answers is not in the document]
Comment (gitizenme) [reply architecture/Mainframe_Data_Access_Framework.md]: Agreed. I'm going to pull your content into the description of a _logical_ environment that relies on container based virtualization.

File: architecture/Mainframe_Data_Access_Framework.md [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+multiple instances of the Engineering environment will be instanced by the
+members of the development team. In the case where the feature or defect is
+simple and does not require integration with other features in the system the
+change is replicated to Staging and is verified. Situations may arise where more
+complex integration and/or testing are required to verify the change as it may
+impact other components or data. To provide for this a QA environment is
+instanced that contains the required set of components that are needed to verify
+the change.
+
+![Environment Diagram](diagrams/Mainframe_Data_Access_Framework-Environment_Diagram.svg)
```
Comment (jeffmaher) [inline architecture/Mainframe_Data_Access_Framework.md]: Clarification question: What do the `www`, `yyy`, `zzz` mean in versions/environment names?

Comment (jeffmaher) [reply architecture/Mainframe_Data_Access_Framework.md]: Question/Suggestion: Does anything (ex: `zzz-final`) need to be appended to version numbers? Could the version number (`v0.0.0`) just be incremented according to the [SemVer guidelines](https://semver.org) without applying a suffix and then an environment points to a given version of the DAF manifest?

File: architecture/Mainframe_Data_Access_Framework.md [hunk 6]
```diff
@@ -0,0 +1,148 @@
...
+
+![Architecture Diagram](diagrams/Mainframe_Data_Access_Framework-Architecture_Diagram.svg)
```
[the remark this answers is not in the document]
Comment (gitizenme) [reply architecture/Mainframe_Data_Access_Framework.md]: Yes, I updated the design to reflect that Marklogic is not required.

File: architecture/Mainframe_Data_Access_Framework.md (hunk 5 above)
Comment (gitizenme) [reply architecture/Mainframe_Data_Access_Framework.md]: This is where I diverge from the SemVer guidelines. There are many OS projects that add the -final extension which reduces ambiguity and makes it more obvious in CI processes. I can go either way on this one.

Comment (gitizenme) [reply architecture/Mainframe_Data_Access_Framework.md]: Missed your first comment, `www`, `yyy`, `zzz` are meant to represent variable version numbers, i.e. placeholders.

Comment (jeffmaher) [reply architecture/Mainframe_Data_Access_Framework.md]: Cool - thanks for the clarification!

**Comment by jeffmaher** [comment]
Looks good to me! Merging

**Comment by jeffmaher** [comment]
Looks good to me! Merging.

**Pull request merged**